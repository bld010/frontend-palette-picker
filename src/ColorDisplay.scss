.ColorDisplay {
  
  & .colors {
    display: flex;
    flex-direction: row;
    justify-content: center;

    & .colorBlock {
      width: 20%;
      height: 40vh;
  
      & .color {
        height: 100%;
        display: flex;
        flex-direction: column;
        justify-content: flex-end;
      }
    }
  }


  & .buttons {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    // padding: 20px;

    & button {
      margin: 10px;
      height: 40px;
      width: 125px;
      // padding-right: 20px;
      border-radius: 8px;
      display: flex;
      flex-direction: row;
      justify-content: space-around;
      align-items: center;
    }
  }

  & .lock {
    color: white;
    opacity: 0.5;
    display: flex;
    flex-direction: row;
    padding: 10px;
    height: 70px;

    & > p {
      color: black;
      background-color: white;
      opacity: 0.5;
      border-radius: 8px;
      font-size: 16px;
      padding: 10px;
      height: 17px;
      margin-top: 20px;
    }
  }
}