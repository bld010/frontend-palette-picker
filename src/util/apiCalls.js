export const getFolders = async () => {
  const url = process.env.REACT_APP_BACKEND_URL + "/api/v1/folders/";
  let res = await fetch(url);
  if (!res.ok) {
    throw new Error("There was an error retrieving your folders!");
  } else {
    let data = await res.json();
    return data;
  }
};

export const getPalettes = async () => {
  const url = process.env.REACT_APP_BACKEND_URL + "/api/v1/palettes/";
  let res = await fetch(url);
  if (!res.ok) {
    throw new Error("There was an error retrieving your palettes!");
  } else {
    let data = await res.json();
    return data;
  }
};

export const deleteFolder = async id => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/${id}`;
  const options = {
    method: "DELETE",
    headers: { "Content-Type": "application/json" }
  };
  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("Cannot delete folder!");
    }
    const deletedFolder = await res.json();
    return deletedFolder;
  } catch (error) {
    throw new Error(error);
  }
};

export const deletePalette = async id => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`;
  const options = {
    method: "DELETE",
    headers: { "Content-Type": "application/json" }
  };
  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("Cannot delete palette!");
    }
    const deletedPalette = await res.json();
    return deletedPalette;
  } catch (error) {
    throw new Error(error);
  }
};

export const postFolder = async folderName => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/`;
  const body = {
    name: folderName
  };
  const options = {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  };

  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("There was an error posting this folder!");
    }
    const newFolder = await res.json();
    return newFolder;
  } catch (error) {
    throw new Error(error);
  }
};

export const postPalette = async (
  colorOne,
  colorTwo,
  colorThree,
  colorFour,
  colorFive,
  folderId,
  paletteName
) => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/`;
  const body = {
    color1: colorOne,
    color2: colorTwo,
    color3: colorThree,
    color4: colorFour,
    color5: colorFive,
    folder_id: folderId,
    name: paletteName
  };
  const options = {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  };

  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("There was an error posting this palette!");
    }
    const newPalette = await res.json();
    return newPalette;
  } catch (error) {
    throw new Error(error);
  }
};

// export const patchFolder = async (folderName, id) => {
//   const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/${id}`;
//   const body = {
//     name: folderName
//   };
//   const options = {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body: JSON.stringify(body)
//   };

//   try {
//     const res = await fetch(url, options);
//     if (!res.ok) {
//       const error = await res.json();
//       throw new Error(error);
//     }
//     const patchedFolder = await res.json();
//     return patchedFolder;
//   } catch (error) {
//     throw new Error(error);
//   }
// };

// export const patchPalette = async (
//   colorOne,
//   colorTwo,
//   colorThree,
//   colorFour,
//   colorFive,
//   folderId,
//   paletteName,
//   id
// ) => {
//   const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`;
//   const body = {
//     color1: colorOne,
//     color2: colorTwo,
//     color3: colorThree,
//     color4: colorFour,
//     color5: colorFive,
//     folder_id: folderId,
//     name: paletteName
//   };
//   const options = {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body: JSON.stringify(body)
//   };

//   try {
//     const res = await fetch(url, options);
//     if (!res.ok) {
//       const error = await res.json();
//       throw new Error(error);
//     }
//     const patchedPalette = await res.json();
//     return patchedPalette;
//   } catch (error) {
//     throw new Error(error);
//   }
// };
