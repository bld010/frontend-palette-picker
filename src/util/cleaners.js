export const cleanPalettes = rawPalettes => {
  return rawPalettes.map(palette => {
    let {
      name,
      colors,
      id,
      folder_id,
      color1,
      color2,
      color3,
      color4,
      color5
    } = palette;

    let cleanedPalette = {
      name,
      id,
      folder_id,
      colors: [color1, color2, color3, color4, color5]
    };

    return cleanedPalette;
  });
};

export const cleanFolders = rawFolders => {
  return rawFolders.map(folder => {
    let { id, name } = folder;
    let cleanedFolder = { id, name, palettes: [] };
    return cleanedFolder;
  });
};

export const cleanData = (folders, palettes) => {
  return folders.map(folder => {
    palettes.forEach(palette => {
      if (palette.folder_id === folder.id) {
        folder.palettes.push(palette);
      }
    });
    return folder;
  });
};
