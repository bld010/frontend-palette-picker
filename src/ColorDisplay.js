import React, { Component } from 'react';
import randomColor from 'randomcolor';
import './ColorDisplay.scss'
import { MdLock, MdLockOpen } from 'react-icons/md'
import { FaSave, FaRandom } from 'react-icons/fa'
import ReactModal from 'react-modal';
import SavePaletteForm from './SavePaletteForm'
import {postPalette} from './util/apiCalls'

export default class ColorDisplay extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentPalette: this.props.palette || null,
      lockedColorIndices: [],
      showModal: false
    }
  }


  componentDidMount = () => {
    if (this.props.palette === null) {
      this.generateNewRandomPalette();
    }
  }
 
  toggleLock = (indexOfClickedColor) => {
    let colors = this.state.currentPalette.colors;
    let name = this.state.currentPalette.name;

    let newColors = colors.map((color, index) => {
      if (indexOfClickedColor === index) {
        color.locked = !color.locked;
        return color
      } else { 
        return color
      }
    })

    this.setState({ currentPalette: {
      name: name,
      colors: newColors
    }})
  }

  generateNewRandomPalette = () => {
    let randomPalette = {
      name: '',
      colors: [
        { hex: randomColor(), locked: false },
        { hex: randomColor(), locked: false },
        { hex: randomColor(), locked: false },
        { hex: randomColor(), locked: false },
        { hex: randomColor(), locked: false }
      ]
    }
    this.setState({ currentPalette: randomPalette })
  }

  getNewColors = () => {
    let colors = this.state.currentPalette.colors;
    let name = this.state.currentPalette.name;

    let newColors = colors.map(color => {
      if (color.locked) {
        return color
      } else {
        return { hex: randomColor(), locked: false }
      }
    })
    this.setState({ currentPalette: {
      name: name,
      colors: newColors
    }})

    this.generateColorsElements();
  }

  generateColorsElements = () => {
    let colors = this.state.currentPalette.colors;

    let colorsElements = colors.map((color, index) => {
      return <div className="colorBlock" key={index}>
        <div className="color"  style={{backgroundColor: color.hex}}>
          <p className="hexCode">{color.hex}</p>
          <div className="lock" onClick={() => this.toggleLock(index)}>
            {color.locked ? <MdLock size={62}/> : <MdLockOpen color="white" size={62}/>}
            {color.locked ? <p className="lockedText">locked</p> : <></>}
          </div>
        </div>
        </div>
    })
    
    return colorsElements;
  }

  displayModal = () => {
    this.setState({showModal: true})
  }

  hideModal = () => {
    this.setState({showModal: false})
  }

  // (colorOne, colorTwo, colorThree, colorFour, colorFive, folderId, paletteName)
  savePalette = async (e, folder, paletteName) => {
    e.preventDefault()
    let colorOne = this.state.currentPalette.colors[0].hex
    let colorTwo = this.state.currentPalette.colors[1].hex
    let colorThree = this.state.currentPalette.colors[2].hex
    let colorFour = this.state.currentPalette.colors[3].hex
    let colorFive = this.state.currentPalette.colors[4].hex
    await postPalette(colorOne, colorTwo, colorThree, colorFour, colorFive, folder.id, paletteName)
    await this.props.reAssignData()
    this.props.displayFolderPalettes(folder.id)
    // this.props.palette()
    this.setState({showModal: false})
  }


  componentDidUpdate = (prevProps) => {
    if (this.props.palette !== prevProps.palette) {
      this.setState( { currentPalette: this.props.palette })
    }
  }



  render = () => {
    if (this.state.currentPalette !== null) {
      let colorsElements = this.generateColorsElements()
      return (
        <div className="ColorDisplay">

          <div className="modal"> 
            <ReactModal isOpen={this.state.showModal} className="ReactModal">
              <SavePaletteForm hideModal={this.hideModal} palette={this.state.currentPalette} savePalette={this.savePalette} folders={this.props.folders} />
            </ReactModal>
          </div>
          <div className="buttons">
            <button onClick={this.getNewColors}><FaRandom size={25} />Random</button>
            <button onClick={this.displayModal}><FaSave size={25} />Save</button>
          </div>
          <div className="colors">
            {colorsElements}
          </div>
            
        </div>
      )
    } else {
      return <></>
    }
  }
}