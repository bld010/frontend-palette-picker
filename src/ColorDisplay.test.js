import React from 'react';
import { shallow } from 'enzyme';
import ColorDisplay from './ColorDisplay';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { postPalette } from './util/apiCalls';

jest.mock('./util/apiCalls')

configure({ adapter: new Adapter() });

describe('ColorDisplay', () => {

  let wrapper;

  let mockPalette = {
    id: 22,
    name: 'Winter Wonders',
    folder_id: 1,
    colors: [{hex: 'blue'}, {hex: 'green'}, {hex: 'pink'}, {hex: 'purple'}, {hex: 'black'}]
  }

  let mockDisplayFolderPalettes = jest.fn();

  let mockReAssignData = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  })

  beforeEach(() => {
    wrapper = shallow(<ColorDisplay 
        palette={mockPalette}
        folders={[]}
        reAssignData={mockReAssignData}
        displayFolderPalettes={mockDisplayFolderPalettes}
      />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('componentDidMount', () => {
    it('should fire generateNewRandomPalette when props.palette is null', () => {
      let wrapper = shallow(<ColorDisplay palette={null} />)
    
      let mockGenerateNewRandomPalette = jest.fn()
      wrapper.instance().generateNewRandomPalette = mockGenerateNewRandomPalette;
    
      wrapper.instance().componentDidMount();
    
      expect(mockGenerateNewRandomPalette).toHaveBeenCalled();
    })

    it('Should not fire generateNewRandomPalette when props.palette is not null', () => {
    
      let mockGenerateNewRandomPalette = jest.fn()
      wrapper.instance().generateNewRandomPalette = mockGenerateNewRandomPalette;
    
      wrapper.instance().componentDidMount();
    
      expect(mockGenerateNewRandomPalette).toHaveBeenCalledTimes(0);
    })
  })

  describe('toggleLock', () => {
    it('should toggle locked property on index of color that was clicked', () => {

      wrapper.instance().setState({ 
        currentPalette: { 
          name: 'Palettey',
          colors: [
          {hex: 'blue', locked:false},
          {hex: 'green', locked: false}
          ]
        }
      })

      let expected = {"colors": [{"hex": "blue", "locked": false}, {"hex": "green", "locked": true}], "name": "Palettey"}

      wrapper.instance().toggleLock(1)

      expect(wrapper.state().currentPalette).toEqual(expected)
    })

  })

  describe('generateNewRandomPalette', () => {
    it('should create an palette with 5 randomly-generated hex codes', () => {
      wrapper.instance().generateNewRandomPalette();

      let colors = wrapper.state().currentPalette.colors

      expect(colors.length).toEqual(5);

      expect(colors[0].hex[0]).toEqual('#');
      expect(colors[0].hex.length).toEqual(7);
    })
  })

  // describe('getNewColors', () => {
  //   it('should generate new colors for any unlocked colors in the palette', () => {

  //   })

  //   it('should fire generateColorsElements', () => {

  //   })
  // })

  describe('displayModal', () => {
    it('should change showModal in state to true', () => {
      wrapper.instance().setState({showModal: false})

      wrapper.instance().displayModal();

      expect(wrapper.state().showModal).toEqual(true);
    })
  })

  describe('hideModal', () => {
    it('should change showModal in state to false', () => {
      wrapper.instance().setState({showModal: true})

      wrapper.instance().hideModal();

      expect(wrapper.state().showModal).toEqual(false);
    })
  })

  describe('savePalette', () => {

    let mockEvent;
    beforeEach(() => {
      mockEvent = {preventDefault: () => {}};

      postPalette.mockImplementation(() => {
        return Promise.resolve()
      })
    })

    it('should call postPalette with current palette info', () => {
      
      wrapper.instance().savePalette(mockEvent, {id: 1}, 'Palletteee');

      expect(postPalette).toHaveBeenCalledWith("blue", "green", "pink", "purple", "black", 1, "Palletteee")
    })

    it('should fire reAssignData', () => {
      
      wrapper.instance().savePalette(mockEvent, {id: 2}, 'Pallettooo');

      expect(mockReAssignData).toHaveBeenCalled();
    })

    it('should fire displayFolderPalettes', () => {

      wrapper.instance().savePalette(mockEvent, {id: 2}, 'Pallettooo');

      expect(mockDisplayFolderPalettes).toHaveBeenCalledWith(2);

    })

    it('should fire hideModal', async () => {

      let mockHideModal = jest.fn()
      wrapper.instance().hideModal = mockHideModal
      wrapper.instance().displayModal();
      await wrapper.instance().savePalette(mockEvent, {id: 2}, 'Pallettooo');
      expect(mockHideModal).toHaveBeenCalled();
    })
  })


  // describe('componentDidUpdate', () => {
  //   it('should set currentPalette in state if palette from props changes', () => {

  //   })
  // })

})