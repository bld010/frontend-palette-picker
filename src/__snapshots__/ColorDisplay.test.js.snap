// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorDisplay should match the snapshot 1`] = `
<div
  className="ColorDisplay"
>
  <div
    className="modal"
  >
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="ReactModal"
      closeTimeoutMS={0}
      isOpen={false}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <SavePaletteForm
        folders={Array []}
        hideModal={[Function]}
        palette={
          Object {
            "colors": Array [
              "blue",
              "green",
              "pink",
              "purple",
              "black",
            ],
            "folder_id": 1,
            "id": 22,
            "name": "Winter Wonders",
          }
        }
        savePalette={[Function]}
      />
    </Modal>
  </div>
  <div
    className="buttons"
  >
    <button
      className="random-button"
      onClick={[Function]}
    >
      <FaRandom
        size={25}
      />
      Random
    </button>
    <button
      className="save-button"
      onClick={[Function]}
    >
      <FaSave
        size={25}
      />
      Save
    </button>
  </div>
  <div
    className="colors"
  >
    <div
      className="colorBlock"
      key="0"
    >
      <div
        className="color"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        <p
          className="hexCode"
        />
        <div
          className="lock"
          onClick={[Function]}
        >
          <MdLockOpen
            color="white"
            size={62}
          />
        </div>
      </div>
    </div>
    <div
      className="colorBlock"
      key="1"
    >
      <div
        className="color"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        <p
          className="hexCode"
        />
        <div
          className="lock"
          onClick={[Function]}
        >
          <MdLockOpen
            color="white"
            size={62}
          />
        </div>
      </div>
    </div>
    <div
      className="colorBlock"
      key="2"
    >
      <div
        className="color"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        <p
          className="hexCode"
        />
        <div
          className="lock"
          onClick={[Function]}
        >
          <MdLockOpen
            color="white"
            size={62}
          />
        </div>
      </div>
    </div>
    <div
      className="colorBlock"
      key="3"
    >
      <div
        className="color"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        <p
          className="hexCode"
        />
        <div
          className="lock"
          onClick={[Function]}
        >
          <MdLockOpen
            color="white"
            size={62}
          />
        </div>
      </div>
    </div>
    <div
      className="colorBlock"
      key="4"
    >
      <div
        className="color"
        style={
          Object {
            "backgroundColor": undefined,
          }
        }
      >
        <p
          className="hexCode"
        />
        <div
          className="lock"
          onClick={[Function]}
        >
          <MdLockOpen
            color="white"
            size={62}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
